python/ray/data/datasource/sql_datasource.py

```diff
@@ -0,0 +1,175 @@
+
+    def create_reader(self, sql: str) -> "Reader":
+        return SQLReader(sql, self.connection_factory)
+
+
+def check_connection_is_dbapi2_compliant(connection) -> None:
+    for attr in "close", "commit", "cursor":
+        if not hasattr(connection, attr):
+            raise ValueError(
+                "Your `connection_factory` created a `Connection` object without a "
+                f"{attr!r} method, but this method is required by the Python DB API2 "
+                "specification. Check that your database connector is DB API2-"
+                "compliant. To learn more, read https://peps.python.org/pep-0249/."
+            )
+
+
+def check_cursor_is_dbapi2_compliant(cursor) -> None:
+    # These aren't all the methods required by the specification, but it's all the ones
+    # we care about.
+    for attr in "execute", "fetchone", "fetchall", "description":
+        if not hasattr(cursor, attr):
+            raise ValueError(
+                "Your database connector created a `Cursor` object without a "
+                f"{attr!r} method, but this method is required by the Python DB API2 "
+                "specification. Check that your database connector is DB API2-"
+                "compliant. To learn more, read https://peps.python.org/pep-0249/."
+            )
+
+
+@contextmanager
+def connect(connection_factory: Callable[[], Connection]) -> Iterator[Cursor]:
+    connection = connection_factory()
+    check_connection_is_dbapi2_compliant(connection)
+
+    try:
+        cursor = connection.cursor()
+        check_cursor_is_dbapi2_compliant(cursor)
+        yield cursor
+
+    finally:
+        # `rollback` is optional since not all databases provide transaction support.
+        try:
+            connection.rollback()
+        except Exception as e:
+            # Each connector implements its own `NotSupportError` class, so we check
+            # the exception's name instead of using `isinstance`.
+            if not (
+                isinstance(e, AttributeError)
+                or e.__class__.__name__ == "NotSupportedError"
+            ):
+                raise e from None
+
+        connection.commit()
+        connection.close()
+
+
+class SQLReader(Reader):
+
+    NUM_SAMPLE_ROWS = 100
+    MIN_ROWS_PER_READ_TASK = 50
+
+    def __init__(self, sql: str, connection_factory: Callable[[], Connection]):
+        self.sql = sql
+        self.connection_factory = connection_factory
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        None
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        # If `parallelism` is 1, directly read the entire database and avoid queries
+        # to fetch a sample block and compute the total number of rows.
+        if parallelism == 1:
+
+            def read_fn() -> Iterable[Block]:
+                with connect(self.connection_factory) as cursor:
+                    cursor.execute(f"SELECT * FROM ({self.sql}) as T")
```

Comment:
why we need to wrap the `self.sql` here?

```
cursor.execute(self.sql)
```

Reply:
We don't -- was an oversight.